mailpoet/lib/Migrations/App/Migration_20240207_105912_App.php

```diff
@@ -0,0 +1,173 @@
+      ->from(ScheduledTaskEntity::class, 't')
+      ->leftJoin('t.subscribers', 's', 'WITH', 's.processed = :unprocessed')
+      ->join('t.sendingQueue', 'q')
+      ->join('q.newsletter', 'n')
+      ->where('t.deletedAt IS NULL')
+      ->andWhere('t.status = :invalid')
+      ->andWhere('s.task IS NULL')
+      ->andWhere('n.deletedAt IS NULL')
+      ->andWhere('n.status = :sending')
+      ->andWhere('n.type IN (:campaignTypes)')
+      ->setParameter('unprocessed', ScheduledTaskSubscriberEntity::STATUS_UNPROCESSED)
+      ->setParameter('invalid', ScheduledTaskEntity::STATUS_INVALID)
+      ->setParameter('sending', NewsletterEntity::STATUS_SENDING)
+      ->setParameter('campaignTypes', NewsletterEntity::CAMPAIGN_TYPES)
+      ->getQuery()
+      ->getSingleColumnResult();
+
+    // update sending queue counts
+    $this->entityManager->createQueryBuilder()
+      ->update(SendingQueueEntity::class, 'q')
+      ->set('q.countProcessed', 'q.countTotal')
+      ->set('q.countToProcess', 0)
+      ->where('q.task IN (:ids)')
+      ->setParameter('ids', $ids)
+      ->getQuery()
+      ->execute();
+
+    // complete the invalid tasks
+    $this->entityManager->createQueryBuilder()
+      ->update(ScheduledTaskEntity::class, 't')
+      ->set('t.status', ':completed')
+      ->where('t.id IN (:ids)')
+      ->setParameter('completed', ScheduledTaskEntity::STATUS_COMPLETED)
+      ->setParameter('ids', $ids)
+      ->getQuery()
+      ->execute();
+
+    // mark newsletters as sent, update "sentAt" (DBAL needed to be able to use JOIN)
+    $newslettersTable = $this->entityManager->getClassMetadata(NewsletterEntity::class)->getTableName();
+    $scheduledTasksTable = $this->entityManager->getClassMetadata(ScheduledTaskEntity::class)->getTableName();
+    $sendingQueuesTable = $this->entityManager->getClassMetadata(SendingQueueEntity::class)->getTableName();
+    $this->entityManager->getConnection()->executeStatement(
+      "
+        UPDATE $newslettersTable n
+        JOIN $sendingQueuesTable q ON n.id = q.newsletter_id
+        JOIN $scheduledTasksTable t ON q.task_id = t.id
+        SET
+          n.status = :sent,
+          n.sent_at = t.updated_at
+        WHERE t.id IN (:ids)
+      ",
+      ['sent' => NewsletterEntity::STATUS_SENT, 'ids' => $ids],
+      ['ids' => Connection::PARAM_INT_ARRAY]
+    );
+  }
+
+  private function backfillMissingDataForMigratedNewsletters(): void {
+    // In https://mailpoet.atlassian.net/browse/MAILPOET-5886 we fixed missing "sent" status
+    // by https://github.com/mailpoet/mailpoet/pull/5416, but didn't backfill missing data.
+
+    // get affected newsletter IDs
+    $ids = $this->entityManager->createQueryBuilder()
+      ->select('n.id')
+      ->from(NewsletterEntity::class, 'n')
+      ->where('n.status = :sent')
+      ->andWhere('n.sentAt IS NULL')
+      ->setParameter('sent', NewsletterEntity::STATUS_SENT)
+      ->getQuery()
+      ->getSingleColumnResult();
```

Comment:
Do you think we need any more filtering here (like only some newsletter types, etc.)?
I suppose that state is invalid, so shouldn't occur, apart from the migrated newsletters.
The original PR had no additional filtering: https://github.com/mailpoet/mailpoet/pull/5416/files

Reply:
No, I think that makes sense.